File: src/main/java/com/wesker926/leetcode/algorithms/p0807/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/12/13
// @description: 模拟遍历
package s1

func maxIncreaseKeepingSkyline(grid [][]int) int {
	maxXs, maxYs := make([]int, len(grid)), make([]int, len(grid))
	for i := range grid {
		for j := range grid[i] {
			maxXs[i] = maxInt(maxXs[i], grid[i][j])
			maxYs[j] = maxInt(maxYs[j], grid[i][j])
		}
	}
	sum, n := 0, 0
	for i := range grid {
		for j := range grid[i] {
			n = maxInt(grid[i][j], minInt(maxXs[i], maxYs[j]))
			sum += n - grid[i][j]
		}
	}
	return sum
}

func maxInt(i, j int) int {
	if i >= j {
		return i
	}
	return j
}

func minInt(i, j int) int {
	if i <= j {
		return i
	}
	return j
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0807/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0807.s1;

/**
 * @author wesker.gh
 * @date 2021/12/13
 * @description 模拟遍历
 */
public class Solution {
    public int maxIncreaseKeepingSkyline(int[][] grid) {
        int n = grid.length;
        int[] maxXs = new int[n], maxYs = new int[n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                maxXs[i] = Math.max(grid[i][j], maxXs[i]);
                maxYs[j] = Math.max(grid[i][j], maxYs[j]);
            }
        }

        int sum = 0, cur;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cur = Math.max(grid[i][j], Math.min(maxXs[i], maxYs[j]));
                sum += cur - grid[i][j];
            }
        }
        return sum;
    }
}
